discovery_engine_core/core/src/state.rs

```diff
@@ -0,0 +1,110 @@
+        #[allow(clippy::cast_possible_truncation)] // bounded by architecture
+        let mut bytes = Vec::with_capacity(size as usize);
+        bytes.push(STATE_VERSION);
+        bincode::serialize_into(&mut bytes, state).map_err(Error::Serialization)?;
+
+        #[cfg(feature = "storage")]
+        {
+            self.storage.state().store(bytes).await?;
+            Ok(Vec::new())
+        }
+
+        #[cfg(not(feature = "storage"))]
+        Ok(bytes)
+    }
+
+    pub(crate) fn deserialize(
+        bytes: &[u8],
+    ) -> Result<(HashMap<Id, Data>, UserInterests, KeyPhrases), Error> {
+        match bytes.get(0) {
+            Some(version) if *version < STATE_VERSION => Ok(Default::default()),
+            Some(&STATE_VERSION) => {
+                bincode::deserialize(&bytes[1..]).map_err(Error::Deserialization)
+            }
+            Some(version) => Err(GenericError::from(format!(
+                "Unsupported serialized data. Found version {} expected {}",
+                *version, STATE_VERSION,
+            ))
+            .into()),
+            None => Err(GenericError::from("Empty serialized data").into()),
+        }
+        // Serialized data could be the partially unversioned data we had before
+        .or_else(|error| {
+            #[derive(Deserialize)]
+            struct SerializedStackState(Vec<u8>);
+            #[derive(Deserialize)]
+            struct SerializedCoiSystemState(Vec<u8>);
+            #[derive(Deserialize)]
+            struct SerializedState {
+                stacks: SerializedStackState,
+                coi: SerializedCoiSystemState,
+            }
+
+            bincode::deserialize::<SerializedState>(bytes)
+                .and_then(|state| {
+                    bincode::deserialize::<HashMap<Id, Data>>(&state.stacks.0)
+                        .map(|stacks| (stacks, state))
+                })
+                .and_then(|(stacks, state)| {
+                    match state.coi.0.get(0) {
+                        Some(&0) => Ok((stacks, UserInterests::default(), KeyPhrases::default())),
+                        Some(&1) => {
+                            #[derive(Deserialize)]
+                            struct CoiSystemState {
+                                user_interests: UserInterests,
+                                key_phrases: KeyPhrases,
+                            }
+
+                            bincode::deserialize::<CoiSystemState>(&state.coi.0[1..])
+                                .map(|coi| (stacks, coi.user_interests, coi.key_phrases))
+                        }
+                        // Serialized data could be the unversioned data we had before
+                        _ => bincode::deserialize::<UserInterests>(bytes)
+                            .map(|user_interests| (stacks, user_interests, KeyPhrases::default())),
+                    }
+                })
+                .map_err(|_| error)
```

Comment:
:thought_balloon: just wondering if it is worth refining the error for some cases e.g. where the `stacks` part was deserialized but not the `coi` part, rather than just returning the error from the level up. perhaps not worth the trouble, just a thought.

Reply:
i'd rather skip this. the code does as much as the current `main` provides. also we will split up state serialization soon anyways, once we persist state more granular in the db.